### PR TITLE
Support `Prefab`

### PR DESCRIPTION
**Pull request by luzhuang**
### Please check if the PR fulfills these requirements

- [x] The commit message follows our [guidelines](https://github.com/galacean/engine/blob/main/.github/COMMIT_MESSAGE_CONVENTION.md)
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

### Other information:

### CODE REVIEW

File: packages/core/src/Entity.ts [hunk 1]
```diff
@@ -98,6 +98,9 @@ export class Entity extends EngineObject {
   /** @internal */
   _isTemplate: boolean = false;
 
+  /** @internal */
+  _isPrefabRoot: boolean = false;
```
Comment (gz65555) [inline packages/core/src/Entity.ts:102]: Is this variable necessary?
I can't see the variable consumed by runtime.

File: packages/core/src/Entity.ts [hunk 2]
```diff
@@ -433,11 +436,13 @@ export class Entity extends EngineObject {
     this._templateResource = templateResource;
   }
 
-  private _createCloneEntity(srcEntity: Entity): Entity {
+  private _createCloneEntity(): Entity {
```
Comment (gz65555) [inline packages/core/src/Entity.ts:439]: Is this modification necessary? or is this related to prefab?
I think you should minimize your modification.

File: packages/core/src/asset/AssetType.ts [hunk 3]
```diff
@@ -30,6 +30,7 @@ export enum AssetType {
   /** AnimatorController. */
   AnimatorController = "AnimatorController",
   /** Prefab.*/
+  Prefab = "Prefab",
   GLTF = "GLTF",
```
Comment (gz65555) [inline packages/core/src/asset/AssetType.ts:34]: should add a comment for gltf

File: packages/loader/src/gltf/GLTFResource.ts [hunk 4]
```diff
@@ -11,13 +11,12 @@ import {
...
 } from "@galacean/engine-core";
+import { PrefabResource } from "../prefab";
 
 /**
  * Product after glTF parser, usually, `defaultSceneRoot` is only needed to use.
  */
-export class GLTFResource extends ReferResource {
-  /** glTF file url. */
-  readonly url: string;
+export class GLTFResource extends PrefabResource {
```
Comment (gz65555) [inline packages/loader/src/gltf/GLTFResource.ts:19]: I can't see many relations between PrefabResource and GLTFResource except a small amount of properties.

File: packages/loader/src/gltf/parser/GLTFParserContext.ts [hunk 5]
```diff
@@ -120,7 +120,7 @@ export class GLTFParserContext {
         const animatorController = glTFResource.animatorController;
 
         if (animatorController) {
-          const animator = glTFResource._defaultSceneRoot.addComponent(Animator);
+          const animator = glTFResource._root.addComponent(Animator);
```
Comment (gz65555) [inline packages/loader/src/gltf/parser/GLTFParserContext.ts:123]: I think you should not modify the files which is related to glTF parser.

File: packages/loader/src/prefab/PrefabParserContext.ts [hunk 6]
```diff
@@ -0,0 +1,9 @@
+import { Engine, Entity } from "@galacean/engine-core";
+import { ParserContext } from "../resource-deserialize/resources/parser/ParserContext";
+import { IHierarchyFile } from "../resource-deserialize";
+
+export class PrefabParserContext extends ParserContext<IHierarchyFile, Entity> {
```
Comment (gz65555) [inline packages/loader/src/prefab/PrefabParserContext.ts:5]: is PrefabParserContext the same as ParserContext?

File: packages/loader/src/resource-deserialize/resources/parser/HierarchyParser.ts [hunk 7]
```diff
@@ -1,12 +1,14 @@
...
+import type { IEntity, IHierarchyFile, IRefEntity, IStrippedEntity } from "../schema";
 import { ReflectionParser } from "./ReflectionParser";
 import { ParserContext } from "./ParserContext";
-import { PrefabParserContext } from "../prefab/PrefabParserContext";
-import { GLTFResource } from "../../../gltf";
+import { PrefabParserContext, PrefabResource } from "../../../prefab";
 
 /** @Internal */
-export default abstract class HierarchyParser<T extends Scene | Entity, V extends ParserContext<IPrefabFile, T>> {
+export default abstract class HierarchyParser<
```
Comment (gz65555) [inline packages/loader/src/resource-deserialize/resources/parser/HierarchyParser.ts:8]: remove export default, use export instead

File: packages/loader/src/prefab/PrefabResource.ts [hunk 8]
```diff
@@ -0,0 +1,32 @@
+import { Engine, Entity, ReferResource } from "@galacean/engine-core";
+
+/**
+ * Product after Prefab parser.
```
Comment (gz65555) [inline packages/loader/src/prefab/PrefabResource.ts:4]: It is very weird about this comment.

File: packages/loader/src/resource-deserialize/resources/parser/HierarchyParser.ts (hunk 7 above)
Comment (luzhuang) [reply packages/loader/src/resource-deserialize/resources/parser/HierarchyParser.ts]: done

File: packages/loader/src/prefab/PrefabResource.ts (hunk 8 above)
Comment (luzhuang) [reply packages/loader/src/prefab/PrefabResource.ts]: done

File: packages/loader/src/prefab/PrefabParserContext.ts (hunk 6 above)
Comment (luzhuang) [reply packages/loader/src/prefab/PrefabParserContext.ts]: removed

File: packages/loader/src/gltf/GLTFResource.ts (hunk 4 above)
Comment (luzhuang) [reply packages/loader/src/gltf/GLTFResource.ts]: reverted

File: packages/loader/src/PrefabLoader.ts [hunk 9]
```diff
@@ -0,0 +1,32 @@
...
+  ResourceManager
+} from "@galacean/engine-core";
+import { PrefabResource } from "./prefab/PrefabResource";
+import { IHierarchyFile, PrefabParser } from "./resource-deserialize";
+
+@resourceLoader(AssetType.Prefab, ["prefab"])
+export class PrefabLoader extends Loader<PrefabResource> {
+  override initialize(_: Engine, configuration: EngineConfiguration): Promise<void> {
+    return Promise.resolve();
+  }
```
Comment (GuoLei1990) [inline packages/loader/src/PrefabLoader.ts:18]: Why override this but do nothing

File: packages/loader/src/gltf/GLTFResource.ts [hunk 10]
```diff
@@ -13,7 +13,7 @@ import {
 } from "@galacean/engine-core";
 
 /**
- * Product after glTF parser, usually, `defaultSceneRoot` is only needed to use.
+ * The glTF asset in runtime, usually, `defaultSceneRoot` is only needed to use.
```
Comment (GuoLei1990) [inline packages/loader/src/gltf/GLTFResource.ts:16]: Opt comments

File: packages/loader/src/prefab/PrefabParser.ts [hunk 11]
```diff
@@ -0,0 +1,36 @@
...
+    super(data, context);
+  }
+
+  protected override handleRootEntity(id: string): void {
+    const rootEntity = this.context.entityMap.get(id);
+    this.prefabResource._root = rootEntity;
+    this.context.entityMap.forEach((entity) => {
+      // @ts-ignore
+      entity._markAsTemplate(this.prefabResource);
+    });
```
Comment (GuoLei1990) [inline packages/loader/src/prefab/PrefabParser.ts:29]: loop outside parse(waste performance)

File: packages/loader/src/prefab/PrefabResource.ts [hunk 12]
```diff
@@ -0,0 +1,32 @@
...
+   * @returns Root entity
+   */
+  instantiate(): Entity {
+    return this._root?.clone();
+  }
+
+  protected override _onDestroy(): void {
+    super._onDestroy();
+    this._root.destroy();
+  }
```
Comment (GuoLei1990) [inline packages/loader/src/prefab/PrefabResource.ts:31]: Not config `_associationSuperResource` and `_disassociationSuperResource`

File: packages/loader/src/prefab/index.ts [hunk 13]
```diff
@@ -0,0 +1,2 @@
+export * from "./PrefabParser";
```
Comment (GuoLei1990) [inline packages/loader/src/prefab/index.ts:1]: Why export `PrefabParser`?

File: packages/loader/src/resource-deserialize/index.ts [hunk 14]
```diff
@@ -6,7 +6,7 @@ import { FileHeader } from "./utils/FileHeader";
 export { MeshDecoder } from "./resources/mesh/MeshDecoder";
 export { Texture2DDecoder } from "./resources/texture2D/TextureDecoder";
 export { ReflectionParser } from "./resources/parser/ReflectionParser";
-export { PrefabParser } from "./resources/prefab/PrefabParser";
+export { PrefabParser } from "../prefab/PrefabParser";
```
Comment (GuoLei1990) [inline packages/loader/src/resource-deserialize/index.ts:9]: export twice?

File: packages/loader/src/resource-deserialize/resources/parser/HierarchyParser.ts [hunk 15]
```diff
@@ -223,27 +221,27 @@ export default abstract class HierarchyParser<T extends Scene | Entity, V extend
...
-          this._traverseAddEntityToMap(entity, context, "");
+          this._generateInstanceContext(entity, instanceContext, "");
 
-          this.prefabContextMap.set(entity, context);
+          this.prefabContextMap.set(entity, instanceContext);
           const cbArray = this.prefabPromiseMap.get(entityConfig.id);
           cbArray &&
             cbArray.forEach((cb) => {
-              cb.resolve(context);
+              cb.resolve(instanceContext);
```
Comment (GuoLei1990) [inline packages/loader/src/resource-deserialize/resources/parser/HierarchyParser.ts:244]: A little waste

File: packages/loader/src/resource-deserialize/index.ts (hunk 14 above)
Comment (luzhuang) [reply packages/loader/src/resource-deserialize/index.ts]: deleted

File: packages/loader/src/prefab/index.ts (hunk 13 above)
Comment (luzhuang) [reply packages/loader/src/prefab/index.ts]: deleted

File: packages/loader/src/PrefabLoader.ts (hunk 9 above)
Comment (luzhuang) [reply packages/loader/src/PrefabLoader.ts:19]: deleted

File: packages/core/src/Entity.ts [hunk 16]
```diff
@@ -433,26 +433,26 @@ export class Entity extends EngineObject {
     this._templateResource = templateResource;
   }
 
-  private _createCloneEntity(srcEntity: Entity): Entity {
-    const cloneEntity = new Entity(srcEntity._engine, srcEntity.name);
+  private _createCloneEntity(): Entity {
```
[the remark this answers is not in the document]
Comment (luzhuang) [reply packages/core/src/Entity.ts:436]: dev/1.3 need merger main

**Review by gz65555 — APPROVED** [review]

**Review by GuoLei1990 — APPROVED** [review]
